### PR TITLE
fix(core): Changes to Native Auth Register Flow

### PR DESCRIPTION
**Pull request by pevey**
# Description

This PR makes changes to the customer registration flow to address two issues:

- Possible email flooding (https://discord.com/channels/1100672177260478564/1229813869548146790)
- Possible data override of existing customers (https://discord.com/channels/1100672177260478564/1228399534745456742)

# Breaking changes

This PR DOES include breaking changes.

The existing sign-up flow checks first to see if a user with that email address exists.  If so, it pulls that data as a starting point.  

- If the existing user is verified, the user record (excluding password) is updated with the submitted data and 'Success' is returned by the mutation.  This is not ideal for two reasons:  The existing data should not be overridden without authentication, and there is no indication to the frontend that the user already exists.  This PR changes this flow to instead return a EmailAddressConflictError.

- If the existing user is not verified, currently a new verification token is generated and emailed to the user upon each sign up attempt.  This could be problematic if a user keeps entering the same (incorrect) email address as it generates spam.  It could also be used maliciously for this effect.  This PR changes the flow to first check to see whether the existing verification token is still valid.  If so, a NotVerifiedError is returned so that the frontend can provide a useful error message and prompt the user to check their email.  If the token is expired, it is possible the record is very old or from someone who initially entered an invalid email address.  For example, joel@example.com accidentally entered joe@example.com.  He realizes his error, corrects his email address, and verifies his joel@example.com address. Later, the real joe@example.com shows up and tries to register.  Only in this case, where the existing token is expired, will a new token be generated.  It will be treated as if the existing user record does not exist.

# Checklist

📌 Always:
- [X] I have set a clear title
- [X] My PR is small and contains a single feature
- [X] I have [checked my own PR](## "Fix typo's and remove unused or commented out code")

👍 Most of the time:
- [ ] I have added or updated test cases
- [ ] I have updated the README if needed

### CODE REVIEW

**Comment by pevey** [comment]
If you're good with this approach, I'll fix the auth-related e2e tests.

**Comment by michaelbromley** [comment]
Hi,

My only concern with the first change is that it opens up the possibility of an account enumeration attack as discussed here:

- https://github.com/vendure-ecommerce/vendure/issues/746

Do you have any thoughts on that?

---

On the second part, I agree with this approach. One point I would change is to not reuse the `NotVerifiedError` since it is semantically a little different:

```gql
"""
Returned if `authOptions.requireVerification` is set to `true` (which is the default)
and an unverified user attempts to authenticate.
"""
type NotVerifiedError implements ErrorResult {
    errorCode: ErrorCode!
    message: String!
}
```

We can create a new error result especially for this case, e.g. `VerificationPendingError` (just a suggestion on the name).

**Comment by pevey** [comment]
On the first topic, it is a good question and one that I also have wondered about.  Security vs UX is always a careful balance. 
 There was discussion of this with the author of the Lucia authentication library in this discord thread if you're interested.  https://discord.com/channels/1004048134218981416/1131337919957839972

In short, he changed my mind about this, and I no longer think feedback about whether an account exists should be hidden from users.  You can see now that pretty much all popular sites give feedback on whether an account exists.  Not doing it is a really poor user experience.  This is applicable to both the sign up and password reset flows.  

To mitigate enumeration, those same sites also use rate-limiting, via actual rate limiting and also bot detection.  Rate-limiting is not something Vendure currently supports out of the box, but it easily could since it is built on NestJS.  That is another PR I'd like to do.  Bot detection is something that everyone hosting ecom should be doing, for this reason but also mostly to protect against carding.  But that leaves the whole industry very reliant on the best bot protection, which is currently Cloudflare with no close second.

So bottom line, I think it is really important to provide the useful error.  Sometimes, you forget whether you created an account somewhere, especially an online store you may not purchase from frequently.  In my checkout flow, if the user is not logged in, I take the email address entered and try to set the customer for order.  If an account exists for that customer, it returns an EmailAddressConflictError.  Which it should, so that you can prompt the user to log in.  This is good UX.  But from this functionality, there already exists the possibility to enumerate accounts without other security protections in place.

**Comment by pevey** [comment]
On the second issues, I did consider adding a new error type, and I'll go ahead and do that if you think it's the best approach.  But an alternative would be to slightly change the wording of the existing error.

**Comment by pevey** [comment]
I'll add the new error and work on the tests.

I'll do a separate PR that implements rate limiting using the NestJS throttler.   https://docs.nestjs.com/security/rate-limiting

**Comment by pevey** [comment]
While we are making changes to the auth flow, I want to bring up another question I have which is very related and should possibly be addressed by this PR.

- As you know, there is a config option for whether to require email verification with the native auth method
- Currently, the logic for adding native auth to a customer account goes like this:
    - If require verification is true, set verified to false, and generate verification token/send notification
    - If require verification is false, set verified to true <- POTENTIAL PROBLEM?
- This logic could be improved, I think.  The account is not really verified.  If we later change the setting to require verification, we have a bunch of accounts with verified set to true, but really they have never been verified.
- In the past, changing this would have broken the sign up flow, but here we are changing the sign up flow already.  Changing it will also break the authentication flow unless we make another small change that only throws the not verified error if the user is not verified AND the require verification is set to true.

Making these small changes would mean:
- The verified status of a user always reflects reality and is not dependent on whether require verification is enabled or disabled. and
- You could switch back and forth between requiring and not requiring verification without major issues.

**Comment by pevey** [comment]
I think I need to generate new type files for the e2e tests, but I'm not sure how to generate them in a way that would include all the types for all plugins so that those tests don't fail.

**Comment by pevey** [comment]
I've hit a wall for now on fixing the e2e tests.  I'm going to take a break from it for now and work on something else and look at this fresh in a couple days.

**Comment by michaelbromley** [comment]
Agreed on the change to verification handling. We'll need to clearly document this change, but it does make sense to do it that way.

Let me know if you need help with the e2e test issues.
